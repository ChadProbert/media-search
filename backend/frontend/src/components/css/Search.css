.searchBar {
	/* search bar width is dependant on the width of the user's screen */
	width: 40% !important;
	/* centers the search bar */
	margin: auto;
}

/* used !important to overwrite bootstrap config */
.searchButton {
	background: rgb(0, 170, 85) !important;
	border: 1px solid rgb(0, 170, 85) !important;
	border-radius: 8px !important;
	box-shadow: 2px 3px 10px 1px rgba(0, 170, 85, 0.363) !important;
}

.searchButton:hover {
	background: rgb(1, 148, 74) !important;
	border: 1px solid rgb(1, 148, 74) !important;
	box-shadow: none !important;
}

.filterSearchBox {
	margin: 0 auto;
	width: 60%;
	padding: 10px;
	background-image: linear-gradient(to right, rgb(0, 170, 85), rgb(0, 130, 85));
	border: 1px solid rgb(0, 170, 85);
	border-radius: 10px;
	color: whitesmoke;
}

.filterHeading {
	font-weight: 600;
	text-transform: uppercase;
}

/* references:

https://www.w3schools.com/cssref/css3_pr_box-shadow.asp

*/
