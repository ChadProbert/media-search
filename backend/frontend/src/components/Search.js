import React, { Component } from "react";
import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import { Container } from "react-grid-system";
import SearchIcon from "@mui/icons-material/Search";

// stylesheets
import "./css/Search.css";

/* takes in a search and filter input from a user and sends a fetch request 
to the iTunes Store API based on the input */
class Search extends Component {
	constructor(props) {
		super(props);

		// initial states
		this.state = { search: "", results: "", type: "" };

		this.handleSearch = this.handleSearch.bind(this);
		this.handleSubmit = this.handleSubmit.bind(this);
		this.handleMediaType = this.handleMediaType.bind(this);
	}

	// captures the user's search input
	handleSearch(e) {
		this.setState({ search: e.target.value });
	}

	// captures the user's filter input
	handleMediaType(e) {
		this.setState({ type: e.target.value });
	}

	async handleSubmit(e) {
		e.preventDefault();
		// fetches data from the iTunes Store API based on the user's input
		await fetch(
			`https://itunes.apple.com/search?term=${this.state.search}&media=${this.state.type}&limit=12`
		)
			.then((res) => res.json())
			// stores the data returned inside the results state
			.then((results) => this.setState({ results: results }))
			/* refreshes the page */
			.then(() => window.location.reload(false))
			.catch((error) => console.log("Error:", error));

		// stores the results data in a config file (search-results.json)
		await fetch("/searches/add", {
			method: "POST",
			headers: {
				"Content-Type": "application/json",
			},
			body: JSON.stringify({
				results: this.state.results,
			}),
		}).catch((error) => console.log("Error", error));
	}

	// creating the UI and adding the relevant event handlers to the user inputs
	/* gave each radio button the media values provided by the Itunes Store API
	documentation */
	// data-test-id "searchTest" (line 65) is used for a frontend unit test
	render() {
		return (
			<Container fluid data-testid="searchTest">
				<Form onSubmit={this.handleSubmit}>
					<Form.Control
						type="text"
						placeholder="Search..."
						onChange={this.handleSearch}
						className="searchBar"
					></Form.Control>
					<br />
					<div className="filterSearchBox">
						<h3 className="filterHeading">Filter</h3>
						<div className="form-check form-check-inline">
							<input
								className="form-check-input"
								type="radio"
								value="movie"
								onClick={this.handleMediaType}
							></input>
							<label className="form-check-label">Movie</label>
						</div>
						<div className="form-check form-check-inline">
							<input
								className="form-check-input"
								type="radio"
								value="podcast"
								onClick={this.handleMediaType}
							></input>
							<label className="form-check-label">Podcast</label>
						</div>
						<div className="form-check form-check-inline">
							<input
								className="form-check-input"
								type="radio"
								value="music"
								onClick={this.handleMediaType}
							></input>
							<label className="form-check-label">Music</label>
						</div>
						<div className="form-check form-check-inline">
							<input
								className="form-check-input"
								type="radio"
								value="musicVideo"
								onClick={this.handleMediaType}
							></input>
							<label className="form-check-label">Music Video</label>
						</div>
						<div className="form-check form-check-inline">
							<input
								className="form-check-input"
								type="radio"
								value="audiobook"
								onClick={this.handleMediaType}
							></input>
							<label className="form-check-label">Audio Book</label>
						</div>
						<div className="form-check form-check-inline">
							<input
								className="form-check-input"
								type="radio"
								value="shortFilm"
								onClick={this.handleMediaType}
							></input>
							<label className="form-check-label">Short Film</label>
						</div>
						<div className="form-check form-check-inline">
							<input
								className="form-check-input"
								type="radio"
								value="tvShow"
								onClick={this.handleMediaType}
							></input>
							<label className="form-check-label">TV Show</label>
						</div>
						<div className="form-check form-check-inline">
							<input
								className="form-check-input"
								type="radio"
								value="software"
								onClick={this.handleMediaType}
							></input>
							<label className="form-check-label">Software</label>
						</div>
						<div className="form-check form-check-inline">
							<input
								className="form-check-input"
								type="radio"
								value="ebook"
								onClick={this.handleMediaType}
							></input>
							<label className="form-check-label">eBook</label>
						</div>
					</div>
					<br />
					<Button type="submit" className="searchButton">
						Search <SearchIcon fontSize="small" />
					</Button>
				</Form>
			</Container>
		);
	}
}

export default Search;

/* references:

https://affiliate.itunes.apple.com/resources/documentation/itunes-store-web-service-search-api/

https://developer.apple.com/library/archive/documentation/AudioVideo/Conceptual/iTuneSearchAPI/Searching.html#//apple_ref/doc/uid/TP40017632-CH5-SW1

https://getbootstrap.com/docs/5.0/forms/checks-radios/

https://react-bootstrap.netlify.app/components/buttons/

https://blog.logrocket.com/a-guide-to-react-onclick-event-handlers-d411943b14dd/

https://mui.com/components/material-icons/

*/
