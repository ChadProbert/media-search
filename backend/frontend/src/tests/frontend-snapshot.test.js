import React from "react";
import renderer from "react-test-renderer";
import Search from "../components/Search.js";

describe("snapshot for Search component", () => {
	test("frontend snapshot", () => {
		/* converts the code inside of the Search component to a json format 
    and is stored as a snapshot inside the project directory */
		const renderedComponent = renderer.create(<Search />).toJSON();
		/* tests to see if the snapshot stored inside the project directory matches 
		with the snapshot taken during the test */
		expect(renderedComponent).toMatchSnapshot();
	});
});

/* references: 

my own work from level 2 task 15

https://www.youtube.com/watch?v=qOaGoujjc3M&ab_channel=SoftwareTestingHelp
 
*/
