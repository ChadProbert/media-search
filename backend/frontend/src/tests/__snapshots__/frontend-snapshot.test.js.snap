// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot for Search component frontend snapshot 1`] = `
<div
  data-testid="searchTest"
  style={
    Object {
      "boxSizing": "border-box",
      "marginLeft": "auto",
      "marginRight": "auto",
      "paddingLeft": 15,
      "paddingRight": 15,
      "position": "relative",
    }
  }
>
  <form
    className=""
    onSubmit={[Function]}
  >
    <input
      className="searchBar form-control"
      onChange={[Function]}
      placeholder="Search..."
      type="text"
    />
    <br />
    <div
      className="filterSearchBox"
    >
      <h3
        className="filterHeading"
      >
        Filter
      </h3>
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          onClick={[Function]}
          type="radio"
          value="movie"
        />
        <label
          className="form-check-label"
        >
          Movie
        </label>
      </div>
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          onClick={[Function]}
          type="radio"
          value="podcast"
        />
        <label
          className="form-check-label"
        >
          Podcast
        </label>
      </div>
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          onClick={[Function]}
          type="radio"
          value="music"
        />
        <label
          className="form-check-label"
        >
          Music
        </label>
      </div>
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          onClick={[Function]}
          type="radio"
          value="musicVideo"
        />
        <label
          className="form-check-label"
        >
          Music Video
        </label>
      </div>
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          onClick={[Function]}
          type="radio"
          value="audiobook"
        />
        <label
          className="form-check-label"
        >
          Audio Book
        </label>
      </div>
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          onClick={[Function]}
          type="radio"
          value="shortFilm"
        />
        <label
          className="form-check-label"
        >
          Short Film
        </label>
      </div>
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          onClick={[Function]}
          type="radio"
          value="tvShow"
        />
        <label
          className="form-check-label"
        >
          TV Show
        </label>
      </div>
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          onClick={[Function]}
          type="radio"
          value="software"
        />
        <label
          className="form-check-label"
        >
          Software
        </label>
      </div>
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          onClick={[Function]}
          type="radio"
          value="ebook"
        />
        <label
          className="form-check-label"
        >
          eBook
        </label>
      </div>
    </div>
    <br />
    <button
      className="searchButton btn btn-primary"
      disabled={false}
      type="submit"
    >
      Search 
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
    </button>
  </form>
  <span
    style={
      Object {
        "clear": "both",
        "display": "table",
      }
    }
  />
</div>
`;
